src/sc/sidechain.cpp

```diff
@@ -375,51 +687,78 @@ bool ScCoinsViewCache::RevertTxOutputs(const CTransaction& tx, int nHeight)
     return true;
 }
 
+bool ScCoinsViewCache::RevertCertOutputs(const CScCertificate& cert, int nHeight)
+{
+    const uint256& scId = cert.scId;
+    const CAmount& totalAmount = cert.totalAmount;
+
+    LogPrint("cert", "%s():%d - removing cert for scId=%s\n", __func__, __LINE__, scId.ToString());
+
+    ScInfo targetScInfo;
+    if (!getScInfo(scId, targetScInfo))
+    {
+        // should not happen
+        LogPrint("cert", "ERROR: %s():%d - scId=%s not in scView\n", __func__, __LINE__, scId.ToString() );
+        return false;
+    }
+
+    targetScInfo.balance += totalAmount;
+    mUpdatedOrNewScInfoList[scId] = targetScInfo;
+
+    LogPrint("cert", "%s():%d - amount restored to scView (amount=%s, resulting bal=%s) %s\n",
+        __func__, __LINE__, FormatMoney(totalAmount), FormatMoney(targetScInfo.balance), scId.ToString());
+
+    return true;
+}
+
 bool ScCoinsViewCache::ApplyMatureBalances(int blockHeight, CBlockUndo& blockundo)
 {
-    LogPrint("sc", "%s():%d - blockHeight=%d, msc_iaundo size=%d\n", __func__, __LINE__, blockHeight,  blockundo.msc_iaundo.size() );
+    LogPrint("sc", "%s():%d - blockHeight=%d, msc_iaundo size=%d\n",
+        __func__, __LINE__, blockHeight,  blockundo.msc_iaundo.size() );
 
     std::set<uint256> allKnowScIds = getScIdSet();
     for(auto it_set = allKnowScIds.begin(); it_set != allKnowScIds.end(); ++it_set)
     {
         const uint256& scId = *it_set;
-        ScInfo info;
-        assert(getScInfo(scId, info));
+        const std::string& scIdString = scId.ToString();
 
-        auto it_ia_map = info.mImmatureAmounts.begin();
+        ScInfo targetScInfo;
+        assert(getScInfo(scId, targetScInfo));
+
+        auto it_ia_map = targetScInfo.mImmatureAmounts.begin();
 
-        while (it_ia_map != info.mImmatureAmounts.end() )
+        while (it_ia_map != targetScInfo.mImmatureAmounts.end() )
```

Comment:
I do not understand why we use the loop at all here.
I believe the whole point of this function is to move immature amount @ height  blockHeight and return error if  blockHeight  is not the first element of the mImmatureAmounts map.
In essence we should never go beyond the first element of the map, so why looping?
Just ```
return false if
targetScInfo.mImmatureAmounts.begin()->first == blockHeight
```
and move immature amounts of targetScInfo.mImmatureAmounts.begin()

Reply:
Ok, since the map is ordered by height, it is enough to have the check just on the first element, others will necessarily have an height grater than the one of the first element.